patches/server/0703-add-per-world-spawn-limits.patch

```diff
@@ -6,43 +6,68 @@ Subject: [PATCH] add per world spawn limits
-@@ -681,6 +681,19 @@ public class PaperWorldConfig {
+@@ -8,6 +8,9 @@ import java.util.function.Function;
+ import java.util.stream.Collector;
+ import java.util.stream.Collectors;
+ 
++import it.unimi.dsi.fastutil.objects.Object2IntMap;
++import it.unimi.dsi.fastutil.objects.Reference2IntMap;
++import it.unimi.dsi.fastutil.objects.Reference2IntOpenHashMap;
+ import net.minecraft.world.entity.MobCategory;
+ import java.util.HashMap;
+ import java.util.List;
+@@ -18,6 +21,7 @@ import net.minecraft.world.entity.EntityType;
+ import net.minecraft.world.entity.monster.Vindicator;
+ import net.minecraft.world.entity.monster.Zombie;
+ import com.destroystokyo.paper.antixray.ChunkPacketBlockControllerAntiXray.EngineMode;
++import net.minecraft.world.level.NaturalSpawner;
+ import org.bukkit.Bukkit;
+ import org.bukkit.configuration.file.YamlConfiguration;
+ import org.spigotmc.SpigotWorldConfig;
+@@ -704,6 +708,25 @@ public class PaperWorldConfig {
          zombieVillagerInfectionChance = getDouble("zombie-villager-infection-chance", zombieVillagerInfectionChance);
      }
  
-+    public int spawnLimitMonsters = -1;
-+    public int spawnLimitAnimals = -1;
-+    public int spawnLimitWaterAnimals = -1;
-+    public int spawnLimitWaterAmbient = -1;
-+    public int spawnLimitAmbient = -1;
++    public Reference2IntMap<MobCategory> perWorldSpawnLimits = new Reference2IntOpenHashMap<>(NaturalSpawner.SPAWNING_CATEGORIES.length);
 +    private void perWorldSpawnLimits() {
-+        spawnLimitMonsters = getInt("spawn-limits.monsters", spawnLimitMonsters);
-+        spawnLimitAnimals = getInt("spawn-limits.animals", spawnLimitAnimals);
-+        spawnLimitWaterAnimals = getInt("spawn-limits.water-animals", spawnLimitWaterAnimals);
-+        spawnLimitWaterAmbient = getInt("spawn-limits.water-ambient", spawnLimitWaterAmbient);
-+        spawnLimitAmbient = getInt("spawn-limits.ambient", spawnLimitAmbient);
++        perWorldSpawnLimits.defaultReturnValue(-1);
++        if (PaperConfig.version < 23) {
++            // ambient category already had correct name
++            perWorldSpawnLimits.put(MobCategory.MONSTER, getInt("spawn-limits.monsters", -1, false));
++            perWorldSpawnLimits.put(MobCategory.CREATURE, getInt("spawn-limits.animals", -1, false));
++            perWorldSpawnLimits.put(MobCategory.WATER_CREATURE, getInt("spawn-limits.water-animals", -1, false));
++            perWorldSpawnLimits.put(MobCategory.WATER_AMBIENT, getInt("spawn-limits.water-ambient", -1, false));
++            set("spawn-limits.monsters", null);
++            set("spawn-limits.animals", null);
++            set("spawn-limits.water-animals", null);
++            set("spawn-limits.water-ambient", null);
++        }
++        for (MobCategory value : NaturalSpawner.SPAWNING_CATEGORIES) {
++            perWorldSpawnLimits.put(value, getInt("spawn-limits." + value.getName(), perWorldSpawnLimits.getInt(value)));
```

Comment:
With this change now the 'spawn-limits.water-animals' has a new name not? (Not sure why spigot call this animals if is creatures)